src/Relude/Nub.hs

```diff
@@ -90,3 +97,27 @@ sortNub = Set.toList . Set.fromList
 -}
 unstableNub :: (Eq a, Hashable a) => [a] -> [a]
 unstableNub = HashSet.toList . HashSet.fromList
+
+{- | A generalized nub for any 'Foldable'. Not necessarily order preserving. -}
+foldNub :: (Foldable t, Ord a) => t a -> Set.Set a
+foldNub = foldr Set.insert Set.empty
+
+{- | A strict variant of 'foldNub' using 'foldl'' instead of 'foldr'-}
+foldNub' :: (Foldable t, Ord a) => t a -> Set.Set a
+foldNub' = foldl' (flip Set.insert) Set.empty
+
+{- | A generalized nub for any 'Foldable' containing 'Hashable' -}
+hashFoldNub :: (Foldable t, Eq a, Hashable a) => t a -> HashSet.HashSet a
+hashFoldNub = foldr HashSet.insert HashSet.empty
+
+{- | A strict variant of 'hashFoldNub' using 'foldl'' instead of 'foldr'-}
+hashFoldNub' :: (Foldable t, Eq a, Hashable a) => t a -> HashSet.HashSet a
+hashFoldNub' = foldl' (flip HashSet.insert) HashSet.empty
+
+{- | A slow @O(n^2)@ nub, with minimal constraints -}
+slowFoldNub :: (Foldable t, Eq a) => t a -> [a]
+slowFoldNub = foldr (\a xs->  if a `elem` xs then xs else a:xs) []
+
+{- | A strict variant of 'slowFoldNub' using 'foldl'' instead of 'foldr' -}
+slowFoldNub' :: (Foldable t, Eq a) => t a -> [a]
+slowFoldNub' = foldl' (\xs a->  if a `elem` xs then xs else a:xs) []
```

Comment:
I'm not sure I want `slowFoldNub` functions 😂 `slowFoldNub` is bad because it searches for the element in non-evaluated tail if the list. `slowFoldNub'` is better but there's `nub`... And if you're working with some specific structure, maybe it's better to use more efficient `nub` for that type? 🤔 For example, `nub` for `Vector` returns `Vector` which is more efficient. And `nub` for `NonEmpty` returns `NonEmpty` which is a stronger type guarantee that just list.

It's an interesting research question, how to abstract and generalise `nub` function. Maybe further research in this direction can be done under separate issue 🙂

Reply:
Well, maybe a better name would be `genericFoldNub`. I agree, it's fairly useless, but I think it's about the only thing you can do since `[a]` is the free monoid of `a`. But yes, not that useful, especially because `Foldable` already requires `Monoid.`

One thing I thought about for the generalization question is whether having a `Nub` typeclass might be interesting:

Maybe something like this?

```haskell
class Nub t a where
  nub :: Eq a => t a -> t a
  nubBy :: (a -> a -> Bool) -> t a -> t a
  ordNub :: Ord a => t a -> t a
  ordNubBy :: (a -> a -> Ordering) -> t a -> t a
  hashNub :: Hashable a => t a -> t a
  hashNubBy :: (a -> Int) -> t a -> t a
```

I think that doesn't work though, because that would mean every instance would have to define all those functions, since they're not derivable from each other.

Actually, couldn't we do

```haskell
class Nub t a where
  nub :: t a -> t a
```
And just have each datatype define its own most efficient instance? Maybe there's even a way to get `deriving Nub` to automatically generate an efficient function based on the other typeclass constraints. That would require an extension though, like Deriving Via, right?

Reply:
I like the solution with:

```haskell
class Nub t a where
  nub :: t a -> t a
```

The open question is how to have both implementations for list with `HashSet` and `Set`. I think that Backpack might be better solution for this problem. I think it's a good idea to try to come up with Backpack interface for `Nub`. But that should go in separate package :slightly_smiling_face:

Reply:
I'm not super familiar with how Backpack works, do you have any recommendations of a good intro I could read? Doing some googling and haven't found anything super clear yet.

Reply:
> do you have any recommendations of a good intro I could read?
hi, there are several ones 
1. [article](https://kowainik.github.io/posts/2018-08-19-picnic-put-containers-into-a-backpack.html)
2. [youtube 1](https://www.youtube.com/watch?v=GLtp3Xy7Rps)
3. [youtube 2](https://www.youtube.com/watch?v=A3ehG4GQpxU)

Reply:
Oh, excellent, thank you! I'll try to study these over the rest of the weekend.